[HDR] Brightness fade in and fade out

mwyrzykowski opened the pull request:
#### 9cb124f4e9b97528c8ee21700e5d10e9f6aa61e1
<pre>
[HDR] Brightness fade in and fade out
<a href="https://bugs.webkit.org/show_bug.cgi?id=283314">https://bugs.webkit.org/show_bug.cgi?id=283314</a>
<a href="https://rdar.apple.com/140139174">rdar://140139174</a>

Reviewed by Simon Fraser.

Update the contentsEDRStrength when the view changes key state
or is otherwise informed by the system.

* Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::updateContentsFormat):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKWindowVisibilityObserver startObserving:]):
(-[WKWindowVisibilityObserver _applicationShouldSuppressHDR:]):
(-[WKWindowVisibilityObserver _applicationShouldAllowHDR:]):
(WebKit::setEDRStrengthRecursive):
(WebKit::setEDRStrength):
(WebKit::WebViewImpl::windowDidBecomeKey):
(WebKit::WebViewImpl::windowDidResignKey):
(WebKit::WebViewImpl::updateHDRState):
(WebKit::WebViewImpl::applicationShouldSuppressHDR):
(WebKit::WebViewImpl::applicationShouldAllowHDR):

Canonical link: <a href="https://commits.webkit.org/291153@main">https://commits.webkit.org/291153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88879b7d8304154a79cac8a7e7b3f622a7d9530f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91988 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96916 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42585 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94038 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11859 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20007 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70577 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28063 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94989 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9030 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50905 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8759 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1063 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41800 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79081 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99036 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19102 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14124 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79667 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78911 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19549 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23368 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12140 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19083 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24292 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18780 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22239 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20531 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
```diff
@@ -814,6 +814,33 @@ - (void)setOwner:(WebCore::PlatformCALayer*)owner
+        BEGIN_BLOCK_OBJC_EXCEPTIONS
+        auto contentsFormat = this->contentsFormat();
+#if ENABLE(PIXEL_FORMAT_RGBA16F)
+        if (contentsFormat == ContentsFormat::RGBA16F) {
```
Nit: `contentsFormat` is outside the #if block, so it could look "unused" on some platforms. Also it's only used once, in the `if` comparison. So I think you could lose it, e.g.: `if (contentsFormat() == ContentsFormat::RGBA16F)`.
(Unless there's another upcoming patch that will use it again?)

squelart reviewed (commented):
Just a C++ nit. LGTM, but I'd prefer someone who fully understands views and tiles to review the logic.

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -369,6 +369,7 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     WEBCORE_EXPORT PluginData& pluginData();
     WEBCORE_EXPORT Ref<PluginData> protectedPluginData();
     void clearPluginData();
+    void allowHDR(bool);
```
This sounds like a getter. It should be `setHDRAllowed` and `isHDRAllowed`.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3174,6 +3175,13 @@ void Page::setActivityState(OptionSet<ActivityState> activityState)
+void Page::allowHDR(bool allow)
+{
+    RefPtr localMainFrame = this->localMainFrame();
+    if (RefPtr view = localMainFrame ? localMainFrame->view() : nullptr)
```
This only does the main frame. We'll need to traverse into subframes.  This will also need extra work for site isolation.

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -1309,6 +1310,9 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     WEBCORE_EXPORT void setPresentingApplicationBundleIdentifier(String&&);
 #endif
 
+    WEBCORE_EXPORT void allowHDR();
```
Super confusing to have two methods called "allowHDR". Also the interaction between "allow" and "suppress" is unclear.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3233,6 +3233,15 @@ void LocalFrameView::updateTiledBackingAdaptiveSizing()
+    if (!tiledBacking)
+        return;
+
+    tiledBacking->allowHDR(allow);
```
This only does the main tiled layer. What about all the other compositing layers, and video?

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -545,12 +545,36 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
 
+    if (properties.changedProperties & LayerChange::AllowHDRChanged) {
+#if ENABLE(PIXEL_FORMAT_RGBA16F)
+        ALLOW_DEPRECATED_DECLARATIONS_BEGIN
```
Share this code with PlatformLayerCocoa.

mwyrzykowski commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -360,6 +360,8 @@ - (void)startObserving:(NSWindow *)window
     [defaultNotificationCenter addObserver:self selector:@selector(_windowWillClose:) name:NSWindowWillCloseNotification object:window];
 
     [defaultNotificationCenter addObserver:self selector:@selector(_screenDidChangeColorSpace:) name:NSScreenColorSpaceDidChangeNotification object:nil];
+    [defaultNotificationCenter addObserver:self selector:@selector(_applicationShouldSuppressHDR:) name:@"NSApplicationShouldBeginSuppressingHighDynamicRangeContentNotification" object:NSApp];
```
cc @squelart

mwyrzykowski committed:
[HDR] Brightness fade in and fade out
https://bugs.webkit.org/show_bug.cgi?id=283314
rdar://140139174

Reviewed by Simon Fraser.

Update the contentsEDRStrength when the view changes key state
or is otherwise informed by the system.

* Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::updateContentsFormat):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKWindowVisibilityObserver startObserving:]):
(-[WKWindowVisibilityObserver _applicationShouldSuppressHDR:]):
(-[WKWindowVisibilityObserver _applicationShouldAllowHDR:]):
(WebKit::setEDRStrengthRecursive):
(WebKit::setEDRStrength):
(WebKit::WebViewImpl::windowDidBecomeKey):
(WebKit::WebViewImpl::windowDidResignKey):
(WebKit::WebViewImpl::updateHDRState):
(WebKit::WebViewImpl::applicationShouldSuppressHDR):
(WebKit::WebViewImpl::applicationShouldAllowHDR):

Canonical link: https://commits.webkit.org/291153@main
